### ISSUE TITLE
JNI not working in executors and URLClassLoader

### ISSUE THREAD

**Issue by niko2014** [SnappyDataInc/snappydata#684]
I need to call jni methods in both driver and executor with `SnappyJob`.

The code works fine in pure spark-jobserver :

```scala

System.loadLibrary("Foo")
Foo.jniMethod1()
val result = snSession.sql("select * from " + table)
result.as[ImageRow].mapPartitions(partition => {
    System.loadLibrary("Foo")
    Foo.jniMethod2(bytes)
})

```

When I submited as a SnappyJob, it ran fine at the first time.
But after that, the exception `libFoo.so already loaded in another classloader` was throwed in executors. But `System.loadLibrary("Foo")` still works fine in driver.
It seems that every submitted job ( closure code of mapPartitions() ) runs in a **new classloader**.

log:

```log
java.lang.UnsatisfiedLinkError: Native Library libFoo.so already loaded in another classloader
        at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1904)
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1854)
        at java.lang.Runtime.loadLibrary0(Runtime.java:870)
        at java.lang.System.loadLibrary(System.java:1122)
        at search34.local.SnappyJobSelectAndSearch$$anonfun$2.apply(SnappyJobSelectAndSearch.scala:105)
        at search34.local.SnappyJobSelectAndSearch$$anonfun$2.apply(SnappyJobSelectAndSearch.scala:91)
        at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:178)
        at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:175)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:804)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:804)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.sql.execution.WholeStageCodegenRDD.compute(WholeStageCodegenExec.scala:496)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:79)
        at org.apache.spark.scheduler.Task.run(Task.scala:107)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:277)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

And I print some classloader info in executors:

```log
# submitted job A
classloader in partitions: java.net.URLClassLoader@21452145
classloader parent in partitions: sun.misc.Launcher$AppClassLoader@7e32c033

# submitted job B
classloader in partitions: java.net.URLClassLoader@4fe93b4e
classloader parent in partitions: sun.misc.Launcher$AppClassLoader@7e32c033
```

**So I remove the `System.loadLibrary("Foo")` code in executor (mapPartitions)**, but `java.lang.UnsatisfiedLinkError: foo.Foo.jniMethod2([B)V` exception was throwed.

tracestack：

```log
org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 8.0 failed 4 times, most recent failure: Lost task 1.3 in stage 8.0 (TID 211, niko-centos7): java.lang.UnsatisfiedLinkError: foo.Foo.jniMethod2([B)V
        at foo.Foo.jniMethod2(Native Method)
        at search34.local.SnappyJobSelectAndSearch$$anonfun$2.apply(SnappyJobSelectAndSearch.scala:110)
        at search34.local.SnappyJobSelectAndSearch$$anonfun$2.apply(SnappyJobSelectAndSearch.scala:91)
        at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:178)
        at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:175)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:804)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:804)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.sql.execution.WholeStageCodegenRDD.compute(WholeStageCodegenExec.scala:496)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:320)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:284)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:79)
        at org.apache.spark.scheduler.Task.run(Task.scala:107)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:277)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

So is there any ways I can get the jni works in executors ( mapPartitions closure ) ?
Thanks.

**Comment by rishitesh** [comment]
@niko2014 You are correct in observing that each job is started with different classloader. For your problem can you keep the loading JNI module a separate class ( inside a static block)  which should be put inside SnappyHome/jars folder in jar format. Then probably you don't have to load from any mapper and only access the library.

**Comment by niko2014** [comment]
@rishitesh  I tried to load library in static block before. But the jar was posted by jobserver `dependent-jar-uris=`. 
I guessed the jar ( with  static loadLibrary ) should be loaded by the **root classloader**. So I put the jni jar to `$SNAPPY_HOME/jars` directory. And The jni register is  **woking**  fine now ! 
I will test again tomorrow and then close the issue. ( Sorry. I tried this solution before actually. While maybe I failed to restart the cluster completely to avoid dirty class-loading state at that time. ) 
Thanks a lot.
